File: src/clients/lib/xmmsclient++/playlist.cpp
/*  XMMS2 - X Music Multiplexer System
 *  Copyright (C) 2003-2016 XMMS2 Team
 *
 *  PLUGINS ARE NOT CONSIDERED TO BE DERIVED WORK !!!
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 */

#include <xmmsclient/xmmsclient.h>
#include <xmmsclient/xmmsclient++/client.h>
#include <xmmsclient/xmmsclient++/playlist.h>
#include <xmmsclient/xmmsclient++/exceptions.h>
#include <xmmsclient/xmmsclient++/signal.h>
#include <xmmsclient/xmmsclient++/helpers.h>
#include <xmmsclient/xmmsclient++/mainloop.h>

#include <boost/bind.hpp>

#include <string>
#include <list>

namespace Xmms
{

	const std::string Playlist::DEFAULT_PLAYLIST = XMMS_ACTIVE_PLAYLIST;


	Playlist::~Playlist()
	{
	}

	StringListResult Playlist::list() const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_list, conn_ ) );
		return StringListResult( res, ml_ );
	}

	VoidResult Playlist::create( const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_create, conn_, playlist.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::load( const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_load, conn_, playlist.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addRecursive( const std::string& url,
	                                   const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_radd, conn_,
		                       playlist.c_str(), url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addRecursiveEncoded( const std::string& url,
	                                          const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_radd_encoded, conn_,
		                       playlist.c_str(), url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addUrl( const std::string& url,
	                             const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_add_url, conn_,
		                       playlist.c_str(), url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addUrl( const std::string& url,
	                             const std::list< std::string >& args,
	                             const std::string& playlist ) const
	{
		std::vector< const char* > cargs;
		fillCharArray( args, cargs );

		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_add_args, conn_,
		                       playlist.c_str(), url.c_str(),
		                       args.size(), &cargs[0] ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addUrlEncoded( const std::string& url,
	                                    const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_add_encoded, conn_,
		                       playlist.c_str(), url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addId( int id,
	                            const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_add_id, conn_,
		                       playlist.c_str(), id ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addIdlist( const Coll::Coll& idlist,
	                                const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_add_idlist, conn_,
		                       playlist.c_str(),
		                       dynamic_cast<const Coll::Idlist&>(idlist).getColl() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::addCollection( const Coll::Coll& collection,
	                                    const std::list< std::string >& order,
	                                    const std::string& playlist ) const
	{
		xmmsv_t *xorder = makeStringList( order );

		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_add_collection, conn_,
		                       playlist.c_str(), collection.getColl(),
		                       xorder ) );

		xmmsv_unref( xorder );

		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::clear( const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_clear, conn_, playlist.c_str() ) );
		return VoidResult( res, ml_ );
	}

	DictResult Playlist::currentPos( const std::string& playlist ) const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playlist_current_pos, conn_,
		                       playlist.c_str() ) );
		return DictResult( res, ml_ );
	}

	StringResult Playlist::currentActive() const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_current_active, conn_ ) );
		return StringResult( res, ml_ );
	}

	VoidResult Playlist::insertUrl( int pos, const std::string& url,
	                                const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_insert_url, conn_,
		                       playlist.c_str(), pos, url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::insertUrl( int pos, const std::string& url,
	                                const std::list< std::string >& args,
	                                const std::string& playlist ) const
	{
		std::vector< const char* > cargs;
		fillCharArray( args, cargs );

		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_insert_args, conn_,
		                       playlist.c_str(), pos, url.c_str(),
		                       args.size(), &cargs[0] ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::insertUrlEncoded( int pos, const std::string& url,
	                                       const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_insert_encoded, conn_,
		                       playlist.c_str(), pos, url.c_str() ) );
		return VoidResult( res, ml_ );
	}


	VoidResult Playlist::insertId( int pos, int id,
	                               const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_insert_id, conn_,
		                       playlist.c_str(), pos, id ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::insertRecursive( int pos, const std::string& url,
	                                      const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_rinsert, conn_,
		                       playlist.c_str(), pos, url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::insertRecursiveEncoded( int pos,
	                                             const std::string& url,
	                                             const std::string& playlist
	                                           ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_rinsert_encoded, conn_,
		                       playlist.c_str(), pos, url.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::insertCollection( int pos, const Coll::Coll& collection,
	                                       const std::list< std::string >& order,
	                                       const std::string& playlist ) const
	{
		xmmsv_t *xorder = makeStringList( order );

		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_insert_collection, conn_,
		                       playlist.c_str(), pos, collection.getColl(),
		                       xorder ) );

		xmmsv_unref( xorder );

		return VoidResult( res, ml_ );
	}

	IntListResult Playlist::listEntries( const std::string& playlist ) const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playlist_list_entries, conn_,
		                       playlist.c_str() ) );
		return IntListResult( res, ml_ );
	}

	VoidResult Playlist::moveEntry( int curpos, int newpos,
	                                const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_move_entry, conn_,
		                       playlist.c_str(), curpos, newpos ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::removeEntry( int pos,
	                                  const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_remove_entry, conn_,
		                       playlist.c_str(), pos ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::remove( const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_remove, conn_, playlist.c_str() ) );
		return VoidResult( res, ml_ );
	}

	IntResult Playlist::setNext( int pos ) const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playlist_set_next, conn_, pos ) );
		return IntResult( res, ml_ );
	}

	IntResult Playlist::setNextRel( signed int pos ) const
	{
		xmmsc_result_t* res = 
		    call( connected_,
		          boost::bind( xmmsc_playlist_set_next_rel, conn_, pos ) );
		return IntResult( res, ml_ );
	}

	VoidResult Playlist::shuffle( const std::string& playlist ) const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_shuffle, conn_, playlist.c_str() ) );
		return VoidResult( res, ml_ );
	}

	VoidResult Playlist::sort( const std::list<std::string>& properties,
	                           const std::string& playlist ) const
	{
		xmmsv_t *xprops = makeStringList( properties );

		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_playlist_sort, conn_,
		                       playlist.c_str(), xprops ) );

		xmmsv_unref( xprops );

		return VoidResult( res, ml_ );
	}

	DictSignal Playlist::broadcastChanged() const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_broadcast_playlist_changed, conn_ ) );
		return DictSignal( res, ml_ );
	}

	DictSignal Playlist::broadcastCurrentPos() const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_broadcast_playlist_current_pos, conn_ ) );
		return DictSignal( res, ml_ );
	}

	StringSignal Playlist::broadcastLoaded() const
	{
		xmmsc_result_t* res =
		    call( connected_,
		          boost::bind( xmmsc_broadcast_playlist_loaded, conn_ ) );
		return StringSignal( res, ml_ );
	}

	Playlist::Playlist( xmmsc_connection_t*& conn, bool& connected,
	                    MainloopInterface*& ml ) :
		conn_( conn ), connected_( connected ), ml_( ml )
	{
	}

}

File: src/clients/lib/java/src/org/xmms2/Playlist.java
/*  XMMS2 - X Music Multiplexer System
 *  Copyright (C) 2003-2007 XMMS2 Team
 *
 *  PLUGINS ARE NOT CONSIDERED TO BE DERIVED WORK !!!
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 *  Lesser General Public License for more details.
 */

package org.xmms2;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.List;
import java.util.ListIterator;

/**
 * Class is usually only used once or on updates replaced by another instance. It holds a list
 * of Titles, the current index, ...
 */
public class Playlist {
	private List pl;
	private HashMap titles;
	private int position = 0;
	private Xmms2 xmms2;
	private int lastAction = -1;
	
	/**
	 * 
	 * @param xmms2
	 */
	protected Playlist(Xmms2 xmms2){
		this.xmms2 = xmms2;
		pl = new ArrayList();
		titles = new HashMap();
		xmms2.playlistListAsync();
		xmms2.playlistGetCurrentIndexAsync();
	}
	
	public int add(long id){
		return xmms2.addId(id);
	}
	
	public int addURL(String url){
		return xmms2.addUrl(url);
	}
	
	public int insertTitle(long id, int index){
		return xmms2.insertId(id, index);
	}
	
	public int insertTitle(String url, int index){
		return xmms2.insertUrl(url, index);
	}
	
	public int move(int source, int dest){
		return xmms2.move(source, dest);
	}
	
	public String toString(){
		String output = "";
		int i = 1;
		for (ListIterator it = pl.listIterator(); it.hasNext();){
			Long t = (Long)it.next();
			output += i++ + "\t" + t + "\n";
		}
		return output;
	}
	
	public ListIterator listIterator(){
		return pl.listIterator();
	}
	
	public Iterator Iterator(){
		return pl.iterator();
	}
	
	/**
	 * @param t
	 * @return		index of Title t
	 */
	public int indexOf(Title t){
		return pl.indexOf(t);
	}
	
	/**
	 * remove Title from playlist
	 * @param t
	 * @return		removed Title
	 */
	public Title removeTitle(Title t){
		xmms2.removeIndex(indexOf(t));
		return t;
	}
	
	/**
	 * Check if this playlist contains the given id
	 * @param id
	 * @return		true if yes, false otherwise
	 */
	public List indicesOfID(long id){
		ArrayList indices = new ArrayList();
		int i = 0;
		for (ListIterator it = pl.listIterator(); it.hasNext();){
			if (((Long)it.next()).longValue() == id) indices.add(new Integer(i));
			i++;
		}
		return indices;
	}
	
	/**
	 * @param index
	 * @return		Title on index
	 */
	public Title titleAt(int index) {
		Object o = titles.get(""+pl.get(index));
		Title t = null;
		if (o instanceof Title)
			t = (Title)o;
		return t;
	}
	
	public long idAt(int index){
		return ((Long)pl.get(index)).longValue();
	}
	
	/**
	 * 
	 * @return		size of Playlist
	 */
	public int length(){
		return pl.size();
	}
	
	/**
	 * 
	 * @return		current position
	 */
	public int getPosition(){
		return position;
	}
	
	/**
	 * @return		current marked Title		
	 */
	public Title getCurrentTitle(){
		return titleAt(position);
	}
	
	/**
	 * set next
	 * @param pos
	 * @return td
	 */
	public int setNextAbs(int pos){
		if (pos > pl.size()-1)
			return xmms2.setNextAbs(pos);
		return xmms2.setNextAbs(pos);
	}
	
	/**
	 * set next
	 * @param pos
	 * @return tid
	 */
	public int setNextRel(int pos){
		if (pos > pl.size()-1)
			return xmms2.setNextRel(pos);
		return xmms2.setNextRel(pos);
	}
	
	/**
	 * clear playlist
	 * @return tid
	 *
	 */
	public int clear(){
		titles.clear();
		return xmms2.clear();
	}
	
	/**
	 * shuffle playlist
	 * @return tid
	 *
	 */
	public int shuffle(){
		return xmms2.shuffle();
	}
	
	/**
	 * sort playlist
	 * @param by 
	 * @return tid
	 *
	 */
	public int sort(String by){
		return xmms2.sort(by);
	}
	
	public int remove(int index){
		return xmms2.removeIndex(index);
	}
	
	public void remove(int indices[]){
		for ( int i = 0; i < indices.length; i++)
			remove(indices[i]-i);
	}
	
	public int getLastAction(){
		return lastAction;
	}
	
	protected void setLastAction(int lastAction){
		this.lastAction = lastAction;
	}
	
	protected void updateList(List l){
		pl = l;
	}
	
	protected void updatePosition(long pos){
		position = (int)pos;
	}
	
	protected void updateTitle(Title t){
		titles.put(""+t.getID(), t);
	}
}
